lib/validation_deso_txn.go

```diff
@@ -0,0 +1,164 @@
+	if txn.TxnMeta == nil {
+		return fmt.Errorf("ValidateDeSoTxnMetadata: Transaction is missing TxnMeta")
+	}
+	if _, err := NewTxnMetadata(txn.TxnMeta.GetTxnType()); err != nil {
+		return errors.Wrapf(err, "ValidateDeSoTxnMetadata: Problem parsing TxnType")
+	}
+	return nil
+}
+
+// ValidateDeSoTxnHash validates that the transaction hash is correctly computed.
+func ValidateDeSoTxnHash(txn *MsgDeSoTxn) error {
+	// Validate transaction hash
+	if txn.Hash() == nil {
+		return fmt.Errorf("ValidateDeSoTxnHash: Problem computing tx hash")
+	}
+	return nil
+}
+
+// ValidateDeSoTxnPublicKey validates that the transaction public key is correctly formatted.
+func ValidateDeSoTxnPublicKey(txn *MsgDeSoTxn) error {
+	// Validate public key
+	if err := IsByteArrayValidPublicKey(txn.PublicKey); err != nil {
+		return errors.Wrapf(err, "ValidateDeSoTxnPublicKey: Problem with public key")
+	}
+	return nil
+}
+
+// ValidateDeSoTxnFormatBalanceModel validates that the transaction is correctly formatted according to the balance model.
+func ValidateDeSoTxnFormatBalanceModel(txn *MsgDeSoTxn, blockHeight uint64, globalParams *GlobalParamsEntry) error {
+	// Validate transaction version
+	if txn.TxnVersion == DeSoTxnVersion0 {
+		return fmt.Errorf("ValidateDeSoTxnFormatBalanceModel: DeSoTxnVersion0 is outdated in balance model")
+	}
+
+	if txn.TxnNonce == nil {
+		return errors.Wrapf(TxErrorNoNonceAfterBalanceModelBlockHeight, "ValidateDeSoTxnFormatBalanceModel: Transaction "+
+			"does not have a nonce.")
+	}
+	if txn.TxnNonce.ExpirationBlockHeight < blockHeight {
+		return errors.Wrapf(TxErrorNonceExpired, "ValidateDeSoTxnFormatBalanceModel: Transaction nonce has expired")
+	}
+	if globalParams.MaxNonceExpirationBlockHeightOffset != 0 &&
+		txn.TxnNonce.ExpirationBlockHeight > blockHeight+globalParams.MaxNonceExpirationBlockHeightOffset {
+		return errors.Wrapf(TxErrorNonceExpirationBlockHeightOffsetExceeded, "ValidateDeSoTxnFormatBalanceModel: Transaction "+
+			"nonce expiration block height offset exceeded")
+	}
+
+	// Verify inputs/outputs.
+	if len(txn.TxInputs) != 0 {
+		return errors.Wrapf(RuleErrorBalanceModelDoesNotUseUTXOInputs, "ValidateDeSoTxnFormatBalanceModel: Balance model "+
+			"transactions should not have any inputs")
+	}
+
+	// Loop through the outputs and do a few sanity checks.
+	var totalOutNanos uint64
+	for _, txout := range txn.TxOutputs {
+		// Check that each output's amount is not bigger than the max as a
+		// sanity check.
+		if txout.AmountNanos > MaxNanos {
+			return errors.Wrapf(RuleErrorOutputExceedsMax, "ValidateDeSoTxnFormatBalanceModel: Output amount %d "+
+				"exceeds max %d", txout.AmountNanos, MaxNanos)
+		}
+		// Check that this output doesn't overflow the total as a sanity
+		// check. This is frankly impossible since our maximum limit is
+		// not close to the max size of a uint64 but check it nevertheless.
+		if _, err := SafeUint64().Add(totalOutNanos, txout.AmountNanos); err != nil {
```

Comment:
I think we want totalOutputNanos on the left side here. Otherwise, we never increase this.